src/sqlfluff/rules/L070.py

```diff
@@ -0,0 +1,249 @@
+"""Implementation of Rule L070."""
+from typing import Optional
+
+from sqlfluff.utils.analysis.select_crawler import Query, SelectCrawler, WildcardInfo
+from sqlfluff.core.rules import BaseRule, LintResult, RuleContext
+from sqlfluff.core.rules.crawlers import SegmentSeekerCrawler
+from sqlfluff.core.rules.doc_decorators import document_groups
+
+
+@document_groups
+class Rule_L070(BaseRule):
+    """Queries within set query produce different numbers of columns.
+
+    **Anti-pattern**
+
+    When writing set expressions
+    , all queries must return the
+    same number of columns
+
+    .. code-block:: sql
+
+        WITH cte AS (
+            SELECT
+                a,
+                b
+            FROM foo
+        )
+        SELECT * FROM cte
+        UNION
+        SELECT
+            c,
+            d,
+            e
+         FROM t
+
+    **Best practice**
+
+    Always specify columns when writing set queries
+    and ensure that they all seleect same number of columns
+
+    .. code-block:: sql
+
+        WITH cte AS (
+            SELECT a, b FROM foo
+        )
+        SELECT
+            a,
+            b
+        FROM cte
+        UNION
+        SELECT
+            c,
+            d
+        FROM t
+    """
+
+    groups = ("all",)
+    crawl_behaviour = SegmentSeekerCrawler({"set_expression"}, provide_raw_stack=True)
+
+    def _find_all_ctes_utils(self, query, cte_dict):
+        """Generate a list of all ctes in a query."""
+        # cte_dict = cte_dict | query.ctes
+        cte_dict.update(query.ctes)
+        for cte_name, cte in query.ctes.items():
+            cte_dict = self._find_all_ctes_utils(cte, cte_dict)
+        return cte_dict
```

Comment:
Nitpick: Since you're updating the dictionary in place, you don't need to return it from the function.